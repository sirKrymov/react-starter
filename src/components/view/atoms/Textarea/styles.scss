@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.textarea {
  position: relative;
  height: 100%;

  textarea {
    -webkit-appearance: none;
    background-color: var(--bgPrimary);
    border-radius: $border-radius;
    transition: border-color $anim-time-fast, background-color $anim-time-fast,
      opacity $anim-time;
    min-height: 130px;
    box-shadow: none;
    padding: 16px;
    outline: none;
    display: block;
    border: 1px solid var(--borderPrimary);
    height: 100%;
    width: 100%;
    color: var(--onPrimary);

    &::placeholder {
      color: var(--textHint);
    }

    &:focus {
      border-color: var(--thirdy);
      box-shadow: none;
    }

    &:read-only {
      border: 1px solid var(--borderPrimary);
    }
  }

  &--error {
    textarea {
      border-color: var(--error);
    }
  }

  &--disable-border {
    textarea {
      background-color: transparent;
      padding-top: 8px;
      border: 0;
    }
  }

  &__length-block {
    position: absolute;
    bottom: -20px;

    &--side-right {
      right: 0;
    }

    &--side-left {
      left: 0;
    }
  }
}
