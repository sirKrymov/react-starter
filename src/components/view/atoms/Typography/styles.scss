// @NOTE: 1rem = 16px by default
@import '../../../../styles/mixins';

.typography {
  &--display-initial {
    display: initial;
  }

  &--display-block {
    display: block;
  }

  &--display-inline {
    display: inline;
  }

  &--variant-inherit {
    font-weight: inherit;
    line-height: inherit;
    font-size: inherit;
  }

  &--variant-subtitle1 {
    font-weight: 500;
    line-height: 1.4;
    font-size: 1.5rem;
  }

  &--variant-subtitle2 {
    font-weight: 500;
    line-height: 1.4;
    font-size: 1.125rem;
  }

  &--variant-subtitle3 {
    font-weight: 500;
    line-height: 1.4;
    font-size: 1rem;
  }

  &--variant-caption {
    line-height: 1.3;
    font-weight: 400;
    font-size: 0.625rem;
  }

  &--variant-body1 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 1.125rem;
  }

  &--variant-body2 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 1rem;
  }

  &--variant-body3 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 0.875rem;
  }

  &--variant-body4 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 12px;
  }

  &--variant-h1 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 2.5rem;
  }

  &--variant-h2 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 2rem;
  }

  &--variant-h3 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 1.5rem;
  }

  &--variant-h4 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 1.125rem;
  }

  &--variant-h5 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 1rem;
  }

  &--variant-h6 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 0.875rem;
  }

  &--color-textSecondary {
    color: var(--textSecondary);
  }

  &--color-textPrimary {
    color: var(--textPrimary);
  }

  &--color-textThirdy {
    color: var(--textThirdy);
  }

  &--color-textLight {
    color: var(--textLight);
  }

  &--color-textDark {
    color: var(--textDark);
  }

  &--color-secondary {
    color: var(--secondary);
  }

  &--color-primary {
    color: var(--primary);
  }

  &--color-thirdy {
    color: var(--thirdy);
  }

  &--color-initial {
    color: initial;
  }

  &--color-inherit {
    color: inherit;
  }

  &--color-valid {
    color: var(--valid);
  }

  &--color-error {
    color: var(--error);
  }

  &--align-left {
    text-align: left;
  }

  &--align-center {
    text-align: center;
  }

  &--align-right {
    text-align: right;
  }

  &--align-justify {
    text-align: justify;
  }

  &--font-roboto {
    font-family: 'Roboto';
  }

  &--uppercase {
    text-transform: uppercase;
  }

  &--nowrap {
    max-width: 100%;
    display: block;
    width: 100%;
    @include text-ellipsis;
  }
}
