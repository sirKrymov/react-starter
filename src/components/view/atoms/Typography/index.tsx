// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactNode, ReactElement } from 'react';

import cn from 'classnames';

import { capitalize } from 'utils';

import './styles.scss';

interface IProps {
  isCapitalize?: boolean;
  component?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'div' | 'p' | 'span';
  paragraph?: boolean;
  uppercase?: boolean;
  className?: string;
  children: ReactNode | string | number;
  variant?:
    | 'subtitle1'
    | 'subtitle2'
    | 'subtitle3'
    | 'inherit'
    | 'caption'
    | 'body1'
    | 'body2'
    | 'body3'
    | 'body4'
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6';
  display?: 'initial' | 'block' | 'inline';
  noWrap?: boolean;
  color?:
    | 'textSecondary'
    | 'textPrimary'
    | 'textThirdy'
    | 'textLight'
    | 'secondary'
    | 'textDark'
    | 'primary'
    | 'initial'
    | 'inherit'
    | 'thirdy'
    | 'error'
    | 'valid';
  align?: 'inherit' | 'left' | 'center' | 'right' | 'justify';
  font?: 'roboto';
}

export function Typography({
  isCapitalize,
  component,
  paragraph,
  uppercase,
  className,
  children,
  variant,
  display = 'initial',
  noWrap,
  color = 'textPrimary',
  align = 'left',
  font = 'roboto'
}: IProps): ReactElement {
  const tagMapping = {
    subtitle1: 'h6',
    subtitle2: 'h6',
    subtitle3: 'h6',
    inherit: 'p',
    body1: 'p',
    body2: 'p',
    body3: 'p',
    body4: 'p',
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6'
  };
  const Tag =
    component || (paragraph ? 'p' : variant && tagMapping[variant]) || 'span';

  if (isCapitalize && typeof children === 'string')
    children = capitalize(children);

  const typographyClass = cn(
    {
      typography: true,
      [`typography--display-${display}`]: display,
      [`typography--variant-${variant}`]: variant,
      [`typography--color-${color}`]: color,
      [`typography--align-${align}`]: align,
      [`typography--font-${font}`]: font,
      [`typography--uppercase`]: uppercase,
      [`typography--nowrap`]: noWrap
    },
    className
  );

  return <Tag className={typographyClass}>{children}</Tag>;
}
