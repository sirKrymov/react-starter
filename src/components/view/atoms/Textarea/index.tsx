// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ChangeEvent, FocusEvent, ReactElement } from 'react';

import cn from 'classnames';

import { Typography } from 'components/view/atoms/Typography';
import { UiSize } from 'components/view/atoms/UiSize';

import './styles.scss';

interface IProps {
  disableBorder?: boolean;
  placeholder?: string;
  lengthSide?: 'left' | 'right';
  className?: string;
  readOnly?: boolean;
  onChange(e: ChangeEvent<any>): void;
  onFocus?(e: FocusEvent<any>): void;
  onBlur?(e: FocusEvent<any>): void;
  height?: 'xs' | 'sm' | 'md' | 'lg' | 'full';
  length?: number;
  width?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full' | 'content';
  error?: boolean;
  type?: 'password' | 'text';
  value: string;
  name: string;
  id: string;
}

export function Textarea({
  disableBorder,
  placeholder,
  lengthSide = 'right',
  className,
  onChange,
  readOnly,
  onFocus,
  onBlur,
  height = 'full',
  length,
  error,
  width = 'full',
  value,
  name,
  id,
  ...rest
}: IProps): ReactElement {
  const textareaClass = cn(
    {
      textarea: true,
      'textarea--disable-border': disableBorder,
      'textarea--error': error
    },
    className
  );
  const lengthClass = cn({
    'textarea__length-block': true,
    [`textarea__length-block--side-${lengthSide}`]: lengthSide
  });

  return (
    <UiSize width={width} height={height}>
      <div className={textareaClass}>
        <textarea
          placeholder={placeholder}
          onChange={onChange}
          readOnly={readOnly}
          onFocus={onFocus}
          onBlur={onBlur}
          value={value}
          name={name}
          id={id}
          {...rest}
        />

        {length && (
          <div className={lengthClass}>
            <Typography variant="body4" color="textLight">
              {value ? value.length : 0} / {length}
            </Typography>
          </div>
        )}
      </div>
    </UiSize>
  );
}
