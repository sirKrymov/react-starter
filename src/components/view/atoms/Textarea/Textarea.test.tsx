import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { Typography } from '../Typography';
import { Textarea } from '../Textarea';

describe('<Textarea> atom component', () => {
  const handleChange = jest.fn();

  let shallowComponent: ShallowWrapper;

  const createComponent = ({
    onChange,
    value,
    ...rest
  }: Record<string, any>) => {
    shallowComponent = shallow(
      <Textarea
        onChange={onChange}
        value={value}
        name="textarea"
        id="check-textarea"
        {...rest}
      />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent({ onChange: handleChange, value: '' });
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });

    it('THEN it should fires on change correctly', () => {
      const event = { target: { value: 'sometext' } };

      expect(shallowComponent.find('textarea').simulate('change', event));
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenLastCalledWith(event);
    });
  });

  describe('WHEN the component have different props', () => {
    it('WHEN to component have length and empty value', () => {
      createComponent({ onChange: handleChange, value: '', length: 20 });

      expect(shallowComponent.find('div.textarea__length-block')).toHaveLength(
        1
      );
      expect(shallowComponent.find(Typography).prop('children')).toEqual([
        0,
        ' / ',
        20
      ]);
    });

    it('WHEN to component have length and value="check"', () => {
      createComponent({ onChange: handleChange, value: 'check', length: 20 });

      expect(shallowComponent.find(Typography).prop('children')).toEqual([
        5,
        ' / ',
        20
      ]);
    });
  });
});
