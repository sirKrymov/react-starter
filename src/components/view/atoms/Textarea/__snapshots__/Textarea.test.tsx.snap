// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<UiSize
  height="full"
  width="full"
>
  <div
    className="textarea"
  >
    <textarea
      id="check-textarea"
      name="textarea"
      onChange={[MockFunction]}
      value=""
    />
  </div>
</UiSize>
`;
